src/codegate/pipeline/pii/pii.py

```diff
@@ -0,0 +1,345 @@
+
+    Methods:
+        name: Returns the name of the pipeline step.
+        _create_chunk: Creates a new ModelResponse chunk with the given content.
+        _format_pii_summary: Formats PII details into a readable summary.
+        process_chunk: Processes a single chunk of stream and adds a notification if PII redacted.
+
+    Attributes:
+        None
+    """
+
+    @property
+    def name(self) -> str:
+        return "pii-redaction-notifier"
+
+    def _create_chunk(self, original_chunk: ModelResponse, content: str) -> ModelResponse:
+        return ModelResponse(
+            id=original_chunk.id,
+            choices=[
+                StreamingChoices(
+                    finish_reason=None,
+                    index=0,
+                    delta=Delta(content=content, role="assistant"),
+                    logprobs=None,
+                )
+            ],
+            created=original_chunk.created,
+            model=original_chunk.model,
+            object="chat.completion.chunk",
+        )
+
+    def _format_pii_summary(self, pii_details: List[Dict[str, Any]]) -> str:
+        """Format PII details into a readable summary"""
+        # Group PII by type
+        pii_types = {}
+        for pii in pii_details:
+            pii_type = pii["type"]
+            if pii_type not in pii_types:
+                pii_types[pii_type] = 0
+            pii_types[pii_type] += 1
+
+        # Format the summary
+        summary_parts = []
+        for pii_type, count in pii_types.items():
+            # Make the type more readable
+            readable_type = pii_type.lower().replace("_", " ")
+            # Add plural 's' if needed
+            if count > 1:
+                summary_parts.append(f"{count} {readable_type}s")
+            else:
+                summary_parts.append(f"{count} {readable_type}")
+        logger.debug(f"PII summary: {summary_parts}")
+        return ", ".join(summary_parts)
+
+    async def process_chunk(
+        self,
+        chunk: ModelResponse,
+        context: OutputPipelineContext,
+        input_context: Optional[PipelineContext] = None,
+    ) -> list[ModelResponse]:
+        """Process a single chunk of the stream"""
+        if (
+            not input_context
+            or not input_context.metadata
+            or input_context.metadata.get("redacted_pii_count", 0) == 0
+        ):
+            return [chunk]
+
+        is_cline_client = any(
+            "Cline" in str(message.trigger_string or "")
+            for message in input_context.alerts_raised or []
+        )
+
+        if len(chunk.choices) > 0 and chunk.choices[0].delta.role:
+            redacted_count = input_context.metadata["redacted_pii_count"]
+            pii_details = input_context.metadata.get("redacted_pii_details", [])
+            pii_summary = self._format_pii_summary(pii_details)
+
+            # Get the redacted text from the metadata
+            redacted_text = input_context.metadata.get("redacted_text", "")
```

Comment:
..but if I understand correctly we don't use the redacted_text during output anywhere?